feat(platform-server): add `provideServerSupport` function to provide server capabilities to an application

alan-agius4 opened the pull request:
This commit adds the `provideServerSupport()` function that returns the `EnvironmentProviders` needed to setup server rendering without NgModules.

eneajaho commented on packages/platform-server/src/provide_server.ts:
```diff
@@ -0,0 +1,38 @@
+ *
+ * @usageNotes
+ *
+ * Basic example of how you a server support to your application:
```
```suggestion
 * Basic example of how you add server support to your application:
```

alan-agius4 committed:
feat(platform-server): add `provideServerSupport` function to provide server capabilities to an application

This commit adds the `provideServerSupport()` function that returns the `EnvironmentProviders` needed to setup server rendering without NgModules.

jessicajaniuk reviewed (approved):
reviewed-for: fw-platform-server, public-api

AndrewKushnir commented on packages/platform-server/src/provide_server.ts:
```diff
@@ -0,0 +1,38 @@
+ */
+export function provideServerSupport(): EnvironmentProviders {
+  return makeEnvironmentProviders([
+    importProvidersFrom(BrowserModule),
+    importProvidersFrom(HttpClientModule),
+    importProvidersFrom(NoopAnimationsModule),
```
We can use standalone APIs here to add providers, WDYT? 
```suggestion
    // importProvidersFrom(BrowserModule), -- do we need it?
    provideHttpClient(withInterceptorsFromDi()),
    provideNoopAnimations(),
```

alan-agius4 replied:
- The `importProvidersFrom(BrowserModule.withServerTransition({appId}))` should not be set in platform-server, it needs to be set on the browser application. This is infact currently incorrect and we should remove it. 

- `withServerTransition` can be deprecated following https://github.com/angular/angular/pull/48253

- `importProvidersFrom(BrowserModule)` or `BROWSER_MODULE_PROVIDERS `, is needed, unless users will provide `importProvidersFrom(BrowserModule)` in their code, as currently `importProvidersFrom(BrowserModule)` is done internally in https://github.com/angular/angular/blob/5e7fc259ead62ee9b4f8a9a77a455065b6a8e2d8/packages/platform-browser/src/browser.ts#L118 where that platform will never be created when using `renderApplication` since this creates another platform.

AndrewKushnir replied:
> The importProvidersFrom(BrowserModule.withServerTransition({appId})) should not be set in platform-server, it needs to be set on the browser application. This is infact currently incorrect and we should remove it. As the APP_ID needs to be set in browser application. Setting it in the renderApplication will cause the APP_ID to mismatch.
>
>withServerTransition can be deprecated following https://github.com/angular/angular/pull/48253 and instead the APP_ID can be provided as a token when needed. By default it can be the comp id of the root component or just ng.

That makes sense 👍 Let's do in a followup PR (plz let me know if you want me to help)!

> importProvidersFrom(BrowserModule) or BROWSER_MODULE_PROVIDERS , is needed, unless users will provide importProvidersFrom(BrowserModule) in their code, as currently importProvidersFrom(BrowserModule) is done internally in

The `BROWSER_MODULE_PROVIDERS` are added to the app providers (not platform providers), so if an app is bootstrapped using `bootstrapApplication`, those providers should be available. For SSR case the code would roughly be like this:

```
renderApplication(() => {
  // The `BROWSER_MODULE_PROVIDERS` are added here
  // (inside the `bootstrapApplication` call).
  return bootstrapApplication(RootCmp, {
    providers: [
      provideServerSupport(),
      // ... the rest of the providers
    ]
  });
});
```

I might be missing something though, please let me know what you think.

alan-agius4 replied:
Thanks @AndrewKushnir, I missed the part that `BROWSER_MODULE_PROVIDERS` are app and not platform providers.

AndrewKushnir reviewed (commented):
@alan-agius4 I think it'd be great to avoid the usage of `importProvidersFrom` and just use the functions that return providers directly.

We could also update the `renderApplication` providers here to use the new method and replace `BrowserModule.withServerTransition` with an `APP_ID` provider. WDYT?

https://github.com/angular/angular/blob/86fc4d3e9825d654c2326afa0af7de6b66632361/packages/platform-server/src/utils.ts#L242-L243

alan-agius4 committed:
fixup! feat(platform-server): add `provideServerSupport` function to provide server capabilities to an application

AndrewKushnir reviewed (approved):
👍

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

alxhub commented:
This PR was merged into the repository by commit 056d68002fbe6024b486bb7220bc77f8f9a07707.
